libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1695,7 +1695,8 @@ string YulUtilFunctions::copyValueArrayStorageToStorageFunction(ArrayType const&
 	solAssert(_fromType.dataStoredIn(DataLocation::Storage) && _toType.baseType()->isValueType(), "");
```

Comment:
Can you add some more assertions, for example that the from base type is a value type?

Reply:
Oh wait a second, in line 1690 we actually assert that from and to are more or less equal. How does this not fail in the tests?

Reply:
Oh wow, looks like this assert is pointless
It ends up in `operator==` of `ReferenceType` which is quite simple:
```
bool operator==(ReferenceType const& _other) const
{
	return location() == _other.location() && isPointer() == _other.isPointer();
}
```
I'll remove this assertion and add more specific once.

Reply:
I added check that both base types are value types, plus that `_fromType.baseType()` is implicitely convertible to `_toType.baseType()`

Reply:
but isn't that operator virtual and overridden by both struct and array?

Reply:
If not we have to change it.

Reply:
It was `dynamic_cast`ed to `ReferenceType`, that is why it was doing the one from `ReferenceType`...
But still , `operator==` from `ArrayType` will not work here as base types are not the same. I have changed assertions completely.

Reply:
I don't understand - it is still virtual - it should always call the most derived, shouldn't it?

Reply:
it wouldn't pass `operator==` in ArrayType, or it has a bug :)
Let me check in debugger what happens with that assertion. Anyway `operator==` for `ArrayType` checks that base types are equal as well. Only way it can work is that for example `byte` and `bytes2` are equal.
Let me check and get back

Reply:
It was calling the one from `ReferenceType` (virtual doesn't work in this case)
```
solAssert(
	*_fromType.copyForLocation(_toType.location(), _toType.isPointer()) == dynamic_cast<ReferenceType const&>(_toType),
	""
);
```

Putting it this way 
```
solAssert(
	_fromType.copyForLocation(_toType.location(), _toType.isPointer()) == &_toType,
	""
);

```
fails, as this one call the one from `ArrayType`